02-ADTs/homework-02/hw-02.hs

```diff
@@ -0,0 +1,198 @@
+of lines to parse, and the log file to parse from (which should also be
+in the same folder as your assignment). For example, after loading
+your assignment into GHCi, type something like this at the prompt:
+testParse parse 10 "error.log"
+Don’t reinvent the wheel! (That’s so last week.) Use Prelude functions to make your solution as concise, high-level, and functional as
+possible. For example, to convert a String like "562" into an Int, you
+can use the read function. Other functions which may (or may not)
+be useful to you include lines, words, unwords, take, drop, and (.).
+-}
+--Exercise 1
+parseMessage :: String -> LogMessage
+parseMessage str = case words str of
+  ("E" : mt : ts : msg) -> LogMessage (Error (read mt)) (read ts) (unwords msg)
+  ("I" : ts : msg)      -> LogMessage Info (read ts) (unwords msg)
+  ("W" : ts : msg)      -> LogMessage Warning (read ts) (unwords msg)
+  _                     -> Unknown str
+
+parse :: String -> [LogMessage]
+parse = map parseMessage . lines
+
+{-
+Unfortunately, due to the error messages being generated by multiple
+servers in multiple locations around the globe, a lightning storm, a
+failed disk, and a bored yet incompetent programmer, the log messages are horribly out of order. Until we do some organizing, there
+will be no way to make sense of what went wrong! We’ve designed a
+data structure that should help—a binary search tree of LogMessages:
+data MessageTree = Leaf
+| Node MessageTree LogMessage MessageTree
+Note that MessageTree is a recursive data type: the Node constructor itself takes two children as arguments, representing the left and
+right subtrees, as well as a LogMessage. Here, Leaf represents the
+empty tree.
+A MessageTree should be sorted by timestamp: that is, the timestamp of a LogMessage in any Node should be greater than all timestamps of any LogMessage in the left subtree, and less than all timestamps of any LogMessage in the right child.
+cis 194: homework 2 4
+Unknown messages should not be stored in a MessageTree since
+they lack a timestamp.
+-}
+
+{-
+Define a function
+insert :: LogMessage -> MessageTree -> MessageTree
+which inserts a new LogMessage into an existing MessageTree, producing a new MessageTree. insert may assume that it is given a
+sorted MessageTree, and must produce a new sorted MessageTree
+containing the new LogMessage in addition to the contents of the
+original MessageTree.
+However, note that if insert is given a LogMessage which is
+Unknown, it should return the MessageTree unchanged.
+-}
+--Exercise 2
+insert :: LogMessage -> MessageTree -> MessageTree
+insert (Unknown _)              tree                        = tree
+insert _                        tree@(Node _ (Unknown _) _) = tree
+insert msg                      Leaf                        = Node Leaf msg Leaf
+insert lm1@(LogMessage _ xts _) (Node left lm2@(LogMessage _ yts _) right)
+  | xts < yts = Node (insert lm1 left) lm2 right
+  | otherwise = Node left lm2 (insert lm1 right)
```

Comment:
This is correct, but I have a few style nitpicks:
- Follow a 4 space indentation style; this is what we use in our production level code. You don't have to make the change for this exercise, but just keep this in mind from now on!
- Try to make your variable names more descriptive. `lm1` is okay for this, but something like `logMsg1` would be better. Just keep this in mind with regards to production code!